### PR TITLE
Add manipulator for global.json

### LINKED ISSUE #10: Add API for adding entries to global.json

### PR DESCRIPTION
**Pull request by tautropfli**

### CODE REVIEW

File: FileManipulator.Test/FileManipulator.Test.csproj [hunk 1]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project Sdk="Microsoft.NET.Sdk; Microsoft.Build.CentralPackageVersions">
+    <PropertyGroup>
+        <TargetFramework>netcoreapp3.1</TargetFramework>
+        <RootNamespace>Messerli.FileManipulator.Test</RootNamespace>
+        <LangVersion>8.0</LangVersion>
+        <Nullable>enable</Nullable>
+        <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+        <WarningsAsErrors />
```
Comment (janhohenheim) [inline FileManipulator.Test/FileManipulator.Test.csproj]: Is this needed?

File: FileManipulator.Test/GlobalJsonManipulatorTest.cs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+                .Build();
+
+        [Theory]
+        [MemberData(nameof(GetModificationData))]
+        public async Task AddsMsBuildSdksToConfig(string expectedConfig, string? existingConfig, GlobalJsonModification modification)
+        {
+            using var testEnvironment = new TestEnvironmentProvider();
+            var filePath = Path.Combine(testEnvironment.RootDirectory, FileName);
+
+            if (existingConfig is { })
```
Comment (janhohenheim) [inline FileManipulator.Test/GlobalJsonManipulatorTest.cs]: Split this into two theories

File: FileManipulator.Test/GlobalJsonManipulatorTest.cs [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+namespace Messerli.FileManipulator.Test
+{
+    public sealed class GlobalJsonManipulatorTest
+    {
+        private const string FileName = "global.json";
+        private const string MsBuildSdkNugetPackageId = "Microsoft.Build.NoTargets";
+        private const string MsBuildSdkVersion = "1.0.0";
+        private static readonly string NewLine = Environment.NewLine;
+
+        private static readonly GlobalJsonModification AddMsBuildSdkModification =
```
Comment (janhohenheim) [inline FileManipulator.Test/GlobalJsonManipulatorTest.cs]: This is a static modification, not a method. Why is it a verb?

File: FileManipulator/FileManipulatorModule.cs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+﻿using Autofac;
+
+namespace Messerli.FileManipulator
+{
+    public class FileManipulatorModule : Module
```
Comment (janhohenheim) [inline FileManipulator/FileManipulatorModule.cs:7]: Add module test if possible

File: FileManipulator/FileManipulatorModule.cs [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+    {
+        protected override void Load(ContainerBuilder builder)
+        {
```
Comment (janhohenheim) [inline FileManipulator/FileManipulatorModule.cs:10]: Register types

File: FileManipulator/GlobalJsonManipulator.cs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+                throw new GlobalJsonManipulationException(exception, filePath);
+            }
+        }
+
+        private static void AddMsBuildSdks(JObject document, IEnumerable<MsBuildSdk> sdks)
+        {
+            var sdkList = document.GetOrInsert(MsBuildSdksJsonProperty, () => new JObject());
+            var sdkListObject = sdkList as JObject
+                ?? throw new MalformedGlobalJsonException(
+                    $"Key '{MsBuildSdksJsonProperty}' is expected to be an object, but is of type {sdkList.Type}");
```
Comment (janhohenheim) [inline FileManipulator/GlobalJsonManipulator.cs]: Clarify that we mean a JSON object

File: FileManipulator/GlobalJsonManipulator.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+            if (existingSdkVersion is { } version && !VersionAreCompatible(version, sdk.Version))
+            {
+                throw new ConflictingMsBuildSdkException(sdk, existingSdkVersion);
+            }
+
+            sdkList[sdk.NuGetPackageId] = sdk.Version;
+        }
+
+        private static bool VersionAreCompatible(string currentVersion, string newVersion)
```
Comment (janhohenheim) [inline FileManipulator/GlobalJsonManipulator.cs]: Versions, in plural

File: FileManipulator/FileManipulator.csproj [hunk 8]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project Sdk="Microsoft.NET.Sdk; Microsoft.Build.CentralPackageVersions">
+  <PropertyGroup>
+    <TargetFramework>netstandard2.1</TargetFramework>
+    <RootNamespace>Messerli.FileManipulator</RootNamespace>
+    <LangVersion>8.0</LangVersion>
+    <Nullable>enable</Nullable>
+    <NullableContextOptions>enable</NullableContextOptions>
+    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <WarningsAsErrors />
```
Comment (janhohenheim) [inline FileManipulator/FileManipulator.csproj]: Use backbone style config globally

File: FileManipulatorAbstractions/GlobalJsonManipulationException.cs [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+    {
+        private readonly string _filePath;
+
+        public GlobalJsonManipulationException(Exception innerException, string filePath)
+            : base(null, innerException)
+        {
+            _filePath = filePath;
+        }
+
+        public override string Message => $"Error manipulating file {_filePath}";
```
Comment (janhohenheim) [inline FileManipulatorAbstractions/GlobalJsonManipulationException.cs]: "... see inner exception for details"

**Review by janhohenheim — APPROVED** [review]

**Review by janhohenheim — CHANGES_REQUESTED** [review]
Approved accidentally

File: FileManipulator.Test/FileManipulator.Test.csproj (hunk 1 above)
Comment (tautropfli) [reply FileManipulator.Test/FileManipulator.Test.csproj]: Already addressed

Comment (tautropfli) [reply FileManipulator.Test/FileManipulator.Test.csproj]: https://github.com/messerli-informatik-ag/meta-generator/pull/14/commits/6c407ef640eae4e4ada1e956055861c135345bde

File: FileManipulator.Test/GlobalJsonManipulatorTest.cs (hunk 2 above)
Comment (tautropfli) [reply FileManipulator.Test/GlobalJsonManipulatorTest.cs]: https://github.com/messerli-informatik-ag/meta-generator/pull/14/commits/28a5dc06d2261711fb152a4e7e14c7d027c46486

File: FileManipulator.Test/GlobalJsonManipulatorTest.cs (hunk 3 above)
Comment (tautropfli) [reply FileManipulator.Test/GlobalJsonManipulatorTest.cs]: "Add" is not intended as a verb here

Comment (tautropfli) [reply FileManipulator.Test/GlobalJsonManipulatorTest.cs]: https://github.com/messerli-informatik-ag/meta-generator/pull/14/commits/5a1213aa05724d351632a9cc108c4581b9b3d180

File: FileManipulator/FileManipulatorModule.cs (hunk 4 above)
Comment (tautropfli) [reply FileManipulator/FileManipulatorModule.cs:7]: https://github.com/messerli-informatik-ag/meta-generator/pull/14/commits/32864bacde278b2b67eb5cc2fd61cb90585d06ff

File: FileManipulator/FileManipulatorModule.cs (hunk 5 above)
Comment (tautropfli) [reply FileManipulator/FileManipulatorModule.cs:10]: https://github.com/messerli-informatik-ag/meta-generator/pull/14/commits/b3e4c85cfd65d3a309eee42308c2d189ef616b4d and https://github.com/messerli-informatik-ag/meta-generator/pull/14/commits/e106256b674f3a6c3e1025196749d5ba00a3890e

File: FileManipulator/GlobalJsonManipulator.cs (hunk 7 above)
Comment (tautropfli) [reply FileManipulator/GlobalJsonManipulator.cs]: https://github.com/messerli-informatik-ag/meta-generator/pull/14/commits/1a1b58cb085ec419f309cf960ae6fa5d665c2893

File: FileManipulator/FileManipulator.csproj (hunk 8 above)
Comment (tautropfli) [reply FileManipulator/FileManipulator.csproj]: https://github.com/messerli-informatik-ag/meta-generator/pull/14/commits/b16cc851c5f310dc754e98ca7625cfeaede53225

File: FileManipulatorAbstractions/GlobalJsonManipulationException.cs (hunk 9 above)
Comment (tautropfli) [reply FileManipulatorAbstractions/GlobalJsonManipulationException.cs]: https://github.com/messerli-informatik-ag/meta-generator/pull/14/commits/d784197f2cd619c703a83ef4dc239092708bed4c

File: FileManipulator/GlobalJsonManipulator.cs (hunk 6 above)
Comment (tautropfli) [reply FileManipulator/GlobalJsonManipulator.cs]: https://github.com/messerli-informatik-ag/meta-generator/pull/14/commits/6869f35d33afa5f80ce9898d03bc12a084638517

**Review by janhohenheim — APPROVED** [review]

**Review by FreeApophis — APPROVED** [review]

**Pull request merged**